0000-macro-value.md

```diff
@@ -0,0 +1,300 @@
+whether `$name` is mentioned in the macro expansion zero, one, or any number times.
+Every expansion of `$name` within the macro expansion refers to the name of the temporary where the captured expression was evaluated.
+If more than one `$:value` capture is present,
+they are evaluated from left to right.
+The intent is that this has identical semantics to that of a function argument capture.
+
+A compiler MAY implement this by expanding to a `match` expression,
+similar to the following:
+
+```rust
+macro_rules! mac! {
+    /* other arms */
+    ( /* other captures */ $name:value /* other captures */ ) => {
+        /* macro body */
+    };
+    /* other arms */
+}
+// "desugars" to
+macro_rules! mac! {
+    /* other arms */
+    ( /* other captures */ $name:expr /* other captures */ ) => {
+        match $name {
+            name => {
+                /* macro body, $name replaced with name (hygienically) */
+            }
+        }
+    };
+    /* other arms */
+}
+```
+
+but the compiler is expected to also handle the case where `$:value` is inside of a macro repetition,
+which cannot be directly implemented by just a desugaring of the `macro_rules!` invocation.
+
+To illustrate (one of) the differences in lifetimes, consider temporary lifetime extension:
+
+```rust
+macro_rules! mlet {
+    ( $e:expr ) => {{
+        let e = $e;
+        e
+    }};
+}
+
+macro_rules! mmatch {
+    ( $e:expr ) => {
+        match $e {
+            e => e,
+        }
+    };
+}
+
+fn mfn<T>(e: T) -> T {
+    e
+}
+
+struct Wrap(&'static str);
+impl Drop for Wrap {
+    fn drop(&mut self) {}
+}
+
+fn a() { let _a = mlet!  (&Wrap("a")); } // borrowck error
+fn b() { let _b = mmatch!(&Wrap("b")); } // compiles
+fn c() { let _c = mfn    (&Wrap("c")); } // compiles
+```
+
+plus there are other differences with drop timing of temporaries,
+especially with macros used in statement position rather than expression position.
+Using `:value` gives the known behavior of a function argument,
+rather than "however you used it."
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+For one, `$:value` is another thing that has to be learned to write effective `macro_rules!` macros.
+However, it replaces the `match` trick, so the author believes this comes out nuetral.
+
+Of course, it expands the language, where a code solution is sufficient, if not elegant.
+
+Additionally,
+this would be the first `macro_rules!` macro where a capture isn't plain token (tree) substitution.
+If the `match` trick is actually fully equivalent to function argument semantics,
+(the author believes this to be true but hasn't exhaustively shown this,)
+then tools such as `cargo expand` (`-Z unpretty`) can use it as the macro expansion.
+But still, this moves `macro_rules!` away from just being token (tree) substitution,
+and gives `macro_rules!` a superpower that isn't available to function-like proc macros.
+
+# Alternatives
```

Comment:
Can this be implemented using a proc macro?

I think the hard constrain on the feature is that the output of `mac!()` using the `value` matcher must be representable as a regular token stream.
(In this sense the "which cannot be directly implemented by just a desugaring of the `macro_rules!` invocation" part sounds a bit concerning.)

But if it can be represented as a token stream, then it should be implementable as a proc macro as well.

Reply:
What this is trying to convey is that `$:value` cannot be implemented as a `macro_rules!` to `macro_rules!` transformation. The `match` trick has the exact semantics that we want here, so the semantics can be gotten by a simple token tree expansion (and thus by proc macros aware of the `match` trick). When I say this would be something `macro_rules!` has that proc macros don't, I'm specifically referring to the `$:value` shorthand, not any expressive power.

As an example,

```rust
macro_rules! sum_twice {
    ( $( $x:value ),+ $(,)? ) => ( 0 $( + $x )* $( + $x)* );
}

sum_twice( $a, $b, $c );
```

has no `macro_rules!` way to implement the desired semantics (due to the repetition), but can potentially expanded as

```rust
match $a { x$a =>
  match $b { x$b =>
    match $c { x$c =>
      0 + x$a + x$b + x$c + x$a + x$b + x$c
    ,}
  ,}
,}
```

(In practice I'm not sure how reasonable actually expanding to this or some other more compact MIR form would be.)